index.js

```diff
@@ -0,0 +1,464 @@
+          s += this.fields[idx].name + ' :: ' + this.fields[idx].type;
+          if (idx === this.fields.length - 1) s += ' ';
+        }
+        return s + '}';
+      },
+      fields: R.map(R.zipObj(['name', 'type']),
+                    R.sortBy(R.head, R.toPairs(fields)))
+    };
+  };
+
+  //  typeEq :: String -> a -> Boolean
+  var typeEq = function(name) {
+    return function(x) {
+      return R.type(x) === name;
+    };
+  };
+
+  //  nullaryTypeFromValue :: t -> Type
+  var nullaryTypeFromValue = function(value) {
+    var name = R.type(value);
+    return NullaryType(name, typeEq(name));
+  };
+
+  //  unaryTypeFromValue :: t a -> (t a -> [a]) -> Type -> Type
+  var unaryTypeFromValue = function(value, _1) {
+    var name = R.type(value);
+    return UnaryType(name, typeEq(name), _1);
+  };
+
+  var _function = /* istanbul ignore next */ function() {};
+
+  $.env = [
+    ($.Array      = unaryTypeFromValue([], R.identity)),
+    ($.Boolean    = nullaryTypeFromValue(false)),
+    ($.Date       = nullaryTypeFromValue(new Date(0))),
+    ($.Error      = nullaryTypeFromValue(new Error())),
+    ($.Function   = nullaryTypeFromValue(_function)),
+    ($.Null       = nullaryTypeFromValue(null)),
+    ($.Number     = nullaryTypeFromValue(0)),
+    ($.RegExp     = nullaryTypeFromValue(/(?:)/)),
+    ($.String     = nullaryTypeFromValue('')),
+    ($.Undefined  = nullaryTypeFromValue(undefined))
+  ];
+
+  //  arity :: Number -> Function -> Function
+  var arity = function(n, f) {
+    return (
+      n === 0 ?
+        function() {
+          return f.apply(this, arguments);
+        } :
+      n === 1 ?
+        function($1) {
+          return f.apply(this, arguments);
+        } :
+      n === 2 ?
+        function($1, $2) {
+          return f.apply(this, arguments);
+        } :
+      n === 3 ?
+        function($1, $2, $3) {
+          return f.apply(this, arguments);
+        } :
+      n === 4 ?
+        function($1, $2, $3, $4) {
+          return f.apply(this, arguments);
+        } :
+      n === 5 ?
+        function($1, $2, $3, $4, $5) {
+          return f.apply(this, arguments);
+        } :
+      n === 6 ?
+        function($1, $2, $3, $4, $5, $6) {
+          return f.apply(this, arguments);
+        } :
+      n === 7 ?
+        function($1, $2, $3, $4, $5, $6, $7) {
+          return f.apply(this, arguments);
+        } :
+      n === 8 ?
+        function($1, $2, $3, $4, $5, $6, $7, $8) {
+          return f.apply(this, arguments);
+        } :
+      // else
+        function($1, $2, $3, $4, $5, $6, $7, $8, $9) {
+          return f.apply(this, arguments);
+        }
+    );
+  };
+
+  //  placeholder :: a -> Boolean
+  var placeholder = function(x) {
+    return x != null && x['@@functional/placeholder'] === true;
+  };
+
+  //  quote :: String -> String
+  var quote = function(s) {
+    return '\u2018' + s + '\u2019';
+  };
+
+  //  forEach :: [a] -> (a -> Number -> Unit) -> Unit
+  var forEach = function(xs, f) {
```

Comment:
Yep! This is to enable some unattractive imperative code, and

```javascript
forEach(xs, function(x) {
  // [snip 50 lines]
});
```

is preferable to

```javascript
forEach(function(x) {
  // [snip 50 lines]
}, xs);
```

since one needn't scroll to the end of the block to see what it is that is being iterated.